### PR TITLE
[AN-5410] Add email screen

### PR DESCRIPTION
**Pull request by JPonte**
#### APK
[Download build #9926](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9926/artifact/build/artifact/wire-dev-PR1321-9926.apk)
[Download build #9927](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9927/artifact/build/artifact/wire-dev-PR1321-9927.apk)
[Download build #9942](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9942/artifact/build/artifact/wire-dev-PR1321-9942.apk)
[Download build #9943](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9943/artifact/build/artifact/wire-dev-PR1321-9943.apk)
[Download build #9944](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9944/artifact/build/artifact/wire-dev-PR1321-9944.apk)
[Download build #9948](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9948/artifact/build/artifact/wire-dev-PR1321-9948.apk)
[Download build #9949](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9949/artifact/build/artifact/wire-dev-PR1321-9949.apk)
[Download build #9974](http://192.168.10.18:8080/job/Pull%20Request%20Builder/9974/artifact/build/artifact/wire-dev-PR1321-9974.apk)
[Download build #10035](http://192.168.10.18:8080/job/Pull%20Request%20Builder/10035/artifact/build/artifact/wire-dev-PR1321-10035.apk)
[Download build #10036](http://192.168.10.18:8080/job/Pull%20Request%20Builder/10036/artifact/build/artifact/wire-dev-PR1321-10036.apk)
[Download build #10060](http://192.168.10.18:8080/job/Pull%20Request%20Builder/10060/artifact/build/artifact/wire-dev-PR1321-10060.apk)
[Download build #10062](http://192.168.10.18:8080/job/Pull%20Request%20Builder/10062/artifact/build/artifact/wire-dev-PR1321-10062.apk)
[Download build #10065](http://192.168.10.18:8080/job/Pull%20Request%20Builder/10065/artifact/build/artifact/wire-dev-PR1321-10065.apk)
[Download build #10565](http://192.168.120.33:8080/job/Pull%20Request%20Builder/10565/artifact/build/artifact/wire-dev-PR1321-10565.apk)
[Download build #10577](http://192.168.120.33:8080/job/Pull%20Request%20Builder/10577/artifact/build/artifact/wire-dev-PR1321-10577.apk)
[Download build #10590](http://192.168.120.33:8080/job/Pull%20Request%20Builder/10590/artifact/build/artifact/wire-dev-PR1321-10590.apk)
[Download build #10608](http://192.168.120.33:8080/job/Pull%20Request%20Builder/10608/artifact/build/artifact/wire-dev-PR1321-10608.apk)

### CODE REVIEW

File: app/src/main/scala/com/waz/zclient/appentry/AppEntryActivity.scala [hunk 1]
```diff
@@ -393,6 +396,12 @@ class AppEntryActivity extends BaseActivity
     enableProgress(false)
   }
 
+  def onShowAddEmail(): Unit = {
+    val transaction: FragmentTransaction = getSupportFragmentManager.beginTransaction
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/AppEntryActivity.scala]: type?

File: app/src/main/scala/com/waz/zclient/appentry/controllers/AppEntryController.scala [hunk 2]
```diff
@@ -304,6 +308,24 @@ class AppEntryController(implicit inj: Injector, eventContext: EventContext) ext
     }
   }
 
+  def addEmail(email: String, password: String): Future[Either[EntryError, Unit]] = {
+    optZms.head.flatMap {
+      case Some(zms) =>
+          zms.account.updateEmail(EmailAddress(email)).future.flatMap {
+            case Right(_) =>
+              zms.accounts.updateCurrentAccount(_.copy(pendingEmail = Some(EmailAddress(email)))).flatMap { _ =>
+                zms.account.updatePassword(password, None)
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/controllers/AppEntryController.scala]: name parameters: `currentPassword = None`

File: app/src/main/scala/com/waz/zclient/appentry/fragments/AddEmailFragment.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    Option(findById[GuidedEditText](getView, R.id.password_field)).foreach { field =>
+      field.setValidator(signInController.passwordValidator)
+      field.setResource(R.layout.guided_edit_text_sign_in__password)
+      field.setText(signInController.password.currentValue.getOrElse(""))
+      field.getEditText.addTextListener(signInController.password ! _)
+    }
+
+    confirmationButton.foreach { button =>
+      button.setOnClickListener(this)
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/fragments/AddEmailFragment.scala]: Use the `RichView#onClick()` util method

File: app/src/main/scala/com/waz/zclient/appentry/fragments/AddEmailFragment.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      button.setAccentColor(Color.WHITE)
+    }
+
+    setConfirmationButtonActive(signInController.isValid.currentValue.getOrElse(false))
+    signInController.isAddEmailValid.onUi { setConfirmationButtonActive }
+    Option(findById[View](R.id.cancel_button)).foreach(_.setOnClickListener(this))
+    Option(findById[View](R.id.ttv_signin_forgot_password)).foreach(_.setOnClickListener(this))
+  }
+
+  private def setConfirmationButtonActive(active: Boolean): Unit =
```
Comment (deanrobertcook) [inline app/src/main/scala/com/waz/zclient/appentry/fragments/AddEmailFragment.scala]: you can have this as a an enclosed `def` within `onCreateView` - then you don't need to keep a member variable of `confirmationButton`

**Comment by someoneEsle** [comment]
Bug 1
I have an account with 7 devices that only has phone number (no email, no password). When logging in, I should be asked to provide an email and password *before* being asked to remove a device (which requires a password)

**Comment by someoneEsle** [comment]
please test

**Comment by someoneEsle** [comment]
Bug 3
Launch the app. Try to log in with fake email and password. Then try to log in with a phone number, type the code. Now you see the "add an email and password" page, but there is already a password (the one you typed in the beginning). I report this bug in many PRs :P

**Comment by someoneEsle** [comment]
Bug 4
On the "add email and password" page, tap back. Endless spinner

**Comment by someoneEsle** [comment]
Bug 7
On "verify email" page, tap back. Endless spinner

Bug 8
On verify email page, tap edit email address, type a new email and password, hit enter. Now you get to the verify email page, but you're also getting an "Invalid information - please verify your details and try again" warning.

**Comment by someoneEsle** [comment]
- [ ] Bug 9
The search UI icon is centered after logging in. There is no helptext, but it should be on the left

**Comment by JPonte** [comment]
build again

**Pull request merged**